### ISSUE TITLE
pg_basebackup task fails when data directory delete in previous task

### ISSUE THREAD

**Issue by iamgini** [samdoran/ansible-role-pgsql-replication#15]
- in `tasks/replica.yml` data directory will be cleaned (delete entire directory) and recreate again.
- but the next step is `Create base backup` which will fail saying no `pg_hba.conf` found (no `postgresql.conf` as well)

```
TASK [samdoran.pgsql_replication : Clear out data directory] ************************************************
changed: [node05-db.lab.local]

TASK [samdoran.pgsql_replication : Create empty data directory] *********************************************
changed: [node05-db.lab.local]

TASK [samdoran.pgsql_replication : Create base backup] ******************************************************
fatal: [node05-db.lab.local]: FAILED! => {"changed": false, "cmd": "pg_basebackup -X stream -D /var/opt/rh/rh-postgresql10/lib/pgsql/data -h 10.6.1.204 -U replicator", "msg": "[Errno 2] No such file or directory", "rc": 2}

```

**Comment by ginigangadharan** [comment]
Suspecting some bug (?) with `become` as well as I tested something below and its working. But not working in role task. (above)

```
- name: test pg_basebackup with sudo
      command: sudo su - postgres pg_basebackup --help 
      become: false
      become_user: postgres
      register: psqlout

    - name: output
      debug:
        msg: "{{ psqlout }}"
```
then,
```
$ ansible-playbook -i inventory psql-test.yaml -u ansible -b
```
and working...
```
TASK [test pg_basebackup with sudo] ***********************************************************************************
[WARNING]: Consider using 'become', 'become_method', and 'become_user' rather than running sudo

changed: [node05-db.lab.local]
TASK [output] ***********************************************************************************************
ok: [node05-db.lab.local] => {
    "msg": {
        "changed": true, 
        "cmd": [
            "sudo", 

````

**Comment by samdoran** [comment]
The paths have changed quite a bit with the change to PostgreSQL 10. I'm working on a fix.

**Comment by samdoran** [comment]
I pushed up some changes that may fix this. If you could test with the latest version of the role and let me know if it's still broken, that would be helpful. I'll look at this some more tomorrow.

**Comment by iamgini** [comment]
@samdoran wonderful !

Sure, I will give a try here. Rebuilding another 5 node test cluster (3+2), so I can try the same role.
I will sure update you..

Thank you

**Comment by iamgini** [comment]
@samdoran 
missed to update. Please see below.

there was an issue with Ansible 2.9.4 as I noticed below thread - https://github.com/ansible/ansible/issues/67245

So upgraded to **2.9.5** (as I saw some fixes on release notes) and see the tasks now. Also added become_flags: `su - postgres -c`  (_Not sure this one really fixed or not_ :))

```
- name: Create base backup
  # edited-here added become_flags: 'su - postgres -c' also make sure ansible version is not 2.9.4 (yet to confirm this)
  shell: "pg_basebackup -X stream -D {{ pgsqlrep_data_path }} -h {{ pgsqlrep_master_address }} -U {{ pgsqlrep_user }}"
  become: yes
  become_user: postgres
  become_flags: 'su - postgres -c'
```

**Other adjustments done**

- Provided pgsql service name manually  (pgsql_version_service : rh-postgresql10-postgresql) as version detection was not working well.
- Removed nginx role from replication playbook – Is that really needed ? not sure any impact due to that.
- Adjusted loop (- name: Add trust in pg_hba.conf)
```
#loop: "{{ pgsqlrep_replica_address }}"
with_items: "{{ pgsqlrep_replica_address }}"
```
**Fail over Playbook**
I have tested fail over using playbook from same repository. (but adjusted pgsql version variables as version detection and data directory are not working by default) 
 
```
vars:
    # edit-here - added pgsql version, path and service name manually
    pgsqlrep_data_path: /var/opt/rh/rh-postgresql10/lib/pgsql/data
    pgsql_version_service : rh-postgresql10-postgresql
    pgsql_bin_path: /opt/rh/rh-postgresql10/root/bin
.
.
.
.
- name: Promote secondary PostgreSQL server to primary
      command: pg_ctl promote #/usr/pgsql-{{ pgsql_version }}/bin/pg_ctl promote
     #edit-here added become parameters
      become: yes
      become_user: postgres
      become_flags: 'su - postgres -c'
      environment:
        PGDATA: "{{ pgsqlrep_data_path }}" #/var/lib/pgsql/{{ pgsql_version }}/data
      ignore_errors: yes
```

**Comment by samdoran** [comment]
I think I have fixed all of these issue now.

> Provided pgsql service name manually (pgsql_version_service : rh-postgresql10-postgresql) as version detection was not working well.

I moved the service name selection out of the filter into OS and version specific vars files.

> Also added become_flags

I think the reason adding `become_flags` to that task fixed the issue is because PostgreSQL 10 on CentOS 7 is coming from software collections. That requires special pathing that is applied in `/etc/profile.d` but only during an interactive login session.

I added the same environment variables to the task that the login script adds, so hopefully the `become_flags` is no longer required.
